### PR TITLE
WIP: [dev-infctx] truncated backprop through time learning & validation experiment

### PR DESCRIPTION
**Pull request by PicoCreator**
- Added `text_rechunk_force` to simplify training using existing HF dataset
- Added `tbptt_learning` to enable truncated backprop through time : https://d2l.ai/chapter_recurrent-neural-networks/bptt.html#truncating-time-steps
- Added notebook, for ease of replication of the trainer validation process
    - I will be adding more notebook examples, for various parts of the training / trainer validation process
- Added `dragon_test.py` , as a means to quickly validate that a trained model works
- Added thread safety for data.py via file locking (required for multi-gpu setup)
    - Added `disable_dataset_preload` parameter, to skip the preload process if needed (eg. readonly filesystem)
- `torch_set_float32_matmul_precision` defaults to `high` (no measured perf penalty)

### CODE REVIEW

File: RWKV-v4neo/src/model.py [hunk 1]
```diff
@@ -462,61 +582,135 @@ def checkpointed_step(idx, targets, mask, prev_loss, last_shift_states,
...
-                loss = L2Wrap.apply(loss, logits, total_mask_sum, submask)
-                new_steps = prev_steps + submask_sum
-                new_loss = prev_loss * (prev_steps / new_steps) + loss * (
-                    1 - prev_steps / new_steps)
-
+            loss = torch.sum(loss * submask) / total_mask_sum  
+            loss = L2Wrap.apply(loss, logits, total_mask_sum, submask)
+            new_steps = prev_steps + submask_sum
+            new_loss = prev_loss + loss
             return new_loss, new_shift_states, new_wkv_states, new_steps
```
Comment (PicoCreator) [inline RWKV-v4neo/src/model.py:629]: Loss here, being added is cumulative - and do not need to be redivided / etc

**Comment by PicoCreator** [comment]
This changes currently freezes on 8xGPU training, gonna debug and see what i can do about it

**Comment by PicoCreator** [comment]
Closed in favour of the splitted up PR's (easier to merge)
